pkg/utils/test/object.go

```diff
@@ -0,0 +1,119 @@
+package test
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/http/httptest"
+	"net/url"
+	"reflect"
+
+	"go.anx.io/go-anxcloud/pkg/api"
+	"go.anx.io/go-anxcloud/pkg/api/types"
+
+	"github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+)
+
+type hookErrorCheck func(context.Context) error
+
+// ObjectTests contains the logic to test any Object implementation with the hooks it implements, checking
+// * if the Object actually implements the interface of the hook
+// * if the Object has an identifier field
+// * calling the hook function with incomplete contexts gives none or the correct error (meaning the error is handled correctly)
+//
+func ObjectTests(o types.Object, hooks ...interface{}) {
+	ginkgo.It("has an identifier", func() {
+		_, err := api.GetObjectIdentifier(o, false)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	})
+
+	for _, hook := range hooks {
+		name := reflect.TypeOf(hook).Elem().Name()
+
+		supportedHooks := map[string]hookErrorCheck{
+			"Object": func(ctx context.Context) error {
+				_, err := o.EndpointURL(ctx)
+				return err
+			},
+			"PaginationSupportHook": func(ctx context.Context) error {
+				_, err := o.(types.PaginationSupportHook).HasPagination(ctx)
+				return err
+			},
+			"RequestBodyHook": func(ctx context.Context) error {
+				_, err := o.(types.RequestBodyHook).FilterAPIRequestBody(ctx)
+				return err
+			},
+			"RequestFilterHook": func(ctx context.Context) error {
+				_, err := o.(types.RequestFilterHook).FilterAPIRequest(ctx, httptest.NewRequest("GET", "/", nil))
+				return err
+			},
+			"ResponseDecodeHook": func(ctx context.Context) error {
+				err := o.(types.ResponseDecodeHook).DecodeAPIResponse(ctx, bytes.NewBuffer([]byte(`{}`)))
+				return err
+			},
+			"ResponseFilterHook": func(ctx context.Context) error {
+				rec := httptest.NewRecorder()
+				rec.WriteHeader(200)
+				_, _ = rec.WriteString(`{}`)
+				_, err := o.(types.ResponseFilterHook).FilterAPIResponse(ctx, rec.Result())
+				return err
+			},
+		}
+
+		if errorCheck, ok := supportedHooks[name]; ok {
```

Comment:
what if `hook` is not in the list of supportedHooks. Do we just not test it then? Or should we at least test whether the hook is actually implemented.

Reply:
good catch, had code organized differently before where this made sense - will change